### ISSUE TITLE
Almost identical concepts (lowercase and uppercase)

### ISSUE THREAD

**Issue by floresbakker** [speced/respec#4821]
### Discussed in https://github.com/speced/respec/discussions/4818

<div type='discussions-op-text'>

<sup>Originally posted by **floresbakker** October 19, 2024</sup>
Hi there ReSpec community!

I am trying to create a specification document that identifies two concepts that have almost identical names. One is for a OWL class called ex:Example and one is for an OWL object property called ex:example. In the RDF/OWL world, it is custom to write classes with a name starting with a upper case letter, while properties are written with a name starting with a lower case letter. The two concepts are semantically different from eachother, hence I want to show this in the ReSpec document. However,  the ReSpec validation shows that the two concepts are seen as one, and it warns me that there is a duplicate definition of ex:example.

    <section>
      <h2>Start your spec!</h2>
      <dfn>[=ex:Example=]</dfn>
      This is a class.      
      <dfn>[=ex:example=]</dfn>
      This is a property.
    </section>

I tried using scoped concepts as described in https://respec.org/docs/#scoped-concepts: 

    <section>
      <h2>Start your spec!</h2>
      <dfn>[=class/ex:Example=]</dfn>
      This is a class.      
      <dfn>[=property/ex:example=]</dfn>
      This is a property.
    </section>

Still I get a duplicate definition warning. 

Even an example that I took from https://respec.org/docs/#scoped-concepts results in a warning:

    <section>
      <h2>Start your spec!</h2>
      <dfn>[=list/for each=]</dfn>
      This is a function for lists.      
      <dfn>[=map/for each=]</dfn>
      This is a function for maps.
    </section>

What am I doing wrong here?
</div>

**Comment by marcoscaceres** [comment]
How we handle this in WebIDL is through a special syntax, like `{{Example/example}}`, which treats things as case insensitive. 

For RDF/OWL, you would need a special syntax to represent and link to RDF/OWL classes and properties. 

This is doable though, if you treat owl as XML. If you want treat it as N3 or whatever notation (I don't know RDF), you would need a specialized parser and to define particular definitions types.

**Comment by floresbakker** [comment]
Anyone that can help? I am trying to come up with a draft report for a W3C community group but this issue is delaying that.

**Comment by floresbakker** [comment]
Could it be that ReSpec’s duplicate detection is missing the scoped term part of the key? In the function collectDfns in 'link-to-dfn.js'. It checks for duplicates here:

```
// check for potential duplicate definition
      if (result.has(dfnFor) && result.get(dfnFor).has(dfnType))
```

Is this unique enough? I am suspecting that the part after '&&' ignores the scope.

**Comment by sidvishnoi** [comment]
Hi @floresbakker,
I think this issue (and the related feature request you filed) can be a major piece of work (and one involving the most sensitive parts of ReSpec), for which I (or other maintainers) do not have time to investigate at present. I also saw your email, but couldn't reply due to busyness. Apologies for the delay this is causing for your report.

Until this is supported, can using IDL instead of concepts work for you? Or trying to define concepts with different terms?
To get your unblocked, can you consider using [Bikeshed](https://github.com/speced/bikeshed) instead if it supports this functionality?

Sending a PR with tests (either as a bug fix, or as a new feature while keeping the backward compatibility) is more than welcome. We can find time to review more easily than investigating and implementing ourselves.

**Comment by tidoust** [comment]
@floresbakker, that's not going to fix your problem with lowercase/uppercase variants but I note you seem to be using `<dfn>[=list/for each=]</dfn>` with a view to defining a term "for each" scoped to "list", and that's not what this syntax does:

- `[=list/for each=]` creates a **reference** to an hypothetical "for each" term scoped to "list". In other words, it creates a `<a>` tag whose inner text is "for each". The "href" attribute would then have the right URL with scope (if that scoped term is actually defined somewhere).
- The wrapping `<dfn>` then creates a **definition** to the term that it finds enclosed. That's "for each" without any scope.

In other words, when you write:

```html
<section>
  <dfn>[=list/for each=]</dfn> blah.
  <dfn>[=map/for each=]</dfn> blah.
</section>
```

... you create two top-level "for each" definitions without any scope. Hence the "duplicate" error that ReSpec reports. To fix this, you need to use the right syntax to define scoped terms as done in [example 25 in the doc](https://respec.org/docs/#scoped-concepts):

```html
<section>
  <dfn data-dfn-for="list">for each</dfn> blah.
  <dfn data-dfn-for="map">for each</dfn> blah.
</section>
```

You can then reference the terms in other parts of the spec through `[=list/for each=]` and `[=map/for each=]`.

Such generic definitions remain case insentive, so again that's not going to solve your core issue. I agree with @marcoscaceres that specific definition types (and syntax) should be used to identify OWL related concepts, as done for CSS, IDL (or CDDL in Bikeshed). That would also make it easier to extract ontology concepts automatically from specs later on (e.g., to produce machine-readable versions of the ontologies).

**Comment by floresbakker** [comment]
@tidoust that does the trick! Many thanks! I misunderstood that example and thought the second part involved some other form of association between concepts and was not part of the scoping functionality. A bit of RTFM if you ask me, although the manual could have been a bit clearer to help silly people like me ;-)

I am generating ReSpec documents for ontologies through my open source linked data based tool OntoReSpec. Not only is the ontology machine-readable (as it is RDF), the document itself is as well machine readable as it is also expressed fully in RDF. Meaning every atom of the document is modeled, including any HTML elements, markdown, definitions, diagrams and more. That resonates with  your statement to extract ontology concepts automatically from specs to produce machine-readable versions of the ontologies, except for the ReSpec document is actually generated based on the ontology & the document itself is machine-readable by design. In addition, OntoReSpec is basically one big computational datamodel itself, expressed in RDF.

Now I am going to adjust OntoReSpec to adopt your helpful tip :) Again, thank you!!

**Comment by floresbakker** [comment]
Again thanks to everyone involved, very much appreciated your efforts! I could finally release ReSpec documentation without errors. See for instance

https://floresbakker.github.io/htmlvoc/
https://minfin-nl.github.io/ITAM/
